### PR TITLE
Add AWS type request signer

### LINKED ISSUE #3: Implement aws-like request signing
See https://github.com/opentelekomcloud/DockerMachineDriver/blob/master/otcgo/signer/signer.go for reference

### PR DESCRIPTION
**Pull request by anton-sidelnikov**
Closes #3

### CODE REVIEW

File: src/oms/core/signer.ts [hunk 1]
```diff
@@ -1,5 +1,9 @@
-import { sha256 } from 'js-sha256';
-import { RequestOpts } from './http';
+/**
+ * Amazon web services (AWS) Signature version 4.
+ * @class Signature
+ */
+
+import * as Crypto from 'crypto-js';
```
Comment (outcatcher) [inline src/oms/core/signer.ts:6]: Please don't use wildcard import if possible

File: src/oms/core/signer.ts [hunk 2]
```diff
@@ -13,34 +17,106 @@ import { RequestOpts } from './http';
...
 //  HexEncode(Hash(RequestPayload))
-function canonicalRequest(config: RequestOpts): string {
-    const url = new URL(config.url, config.baseURL)
-    const queryString = url.search.substring(1)
-    const canonHeaders = Object.entries(config.headers).map(e => `${e[0].toLowerCase()}:${e[1].trim()}\n`).sort().join('')
-    const signedHeaders = Object.keys(config.headers).sort((a: string, b: string): number => (a[0].toLowerCase() < b[0].toLowerCase() ? -1 : 1)).join(';')
-    const hash = hexEncode(sha256(config.body!))
-    return `${config.method}\n${url.pathname}\n${queryString}\n${canonHeaders}\n${signedHeaders}\n${hash}`
+
+export class SignatureInputData{
```
Comment (outcatcher) [inline src/oms/core/signer.ts:21]: Class constructor is never used, can it be an interface?

File: src/oms/core/signer.ts [hunk 3]
```diff
@@ -13,34 +17,106 @@ import { RequestOpts } from './http';
 //  CanonicalHeaders + '\n' +
```
Comment (outcatcher) [inline src/oms/core/signer.ts:17]: Should this comment be here?

**Review by outcatcher — CHANGES_REQUESTED** [review]
Please also add a functional test for authenticating with AK/SK

File: src/oms/core/signer.ts (hunk 3 above)
Comment (anton-sidelnikov) [reply src/oms/core/signer.ts:17]: no

File: src/oms/core/signer.ts [hunk 4]
```diff
@@ -13,34 +17,106 @@ import { RequestOpts } from './http';
...
+    }
+
+    private signString(input: SignatureInputData, dateStamp: string, stringToSign: string) {
+        const signingKey = Signature.getSignatureKey(input.secretKey, dateStamp, input.region, input.service);
+        const signature = Crypto.HmacSHA256(stringToSign, signingKey).toString();
+        return signature;
+    }
+
+    private static generateAuthorizationHeader(algorithm: string, input: SignatureInputData, credentialScope: string, signedHeaders: string, signature: any) {
+        return algorithm + ' ' + 'Credential=' + input.accessKey + '/'
```
Comment (lego963) [inline src/oms/core/signer.ts:105]: `'${algorithm } Credential=${input.accessKey }/${credentialScope }, SignedHeaders=${signedHeaders}, Signature=${signature}'`

Comment (lego963) [reply src/oms/core/signer.ts:105]: https://basarat.gitbook.io/typescript/future-javascript/template-strings#string-interpolation

File: src/oms/core/signer.ts [hunk 5]
```diff
@@ -1,46 +1,102 @@
...
+
+    private static generateAuthorizationHeader(algorithm: string, input: SignatureInputData, credentialScope: string, signedHeaders: string, signature: any) {
+        return `${algorithm} Credential=${input.accessKey}/${credentialScope}, SignedHeaders=${signedHeaders}, Signature=${signature}`
+    }
+
+    private static getSignatureKey(
+        key: string,
+        dateStamp: string,
+        regionName: string,
+        serviceName: string): any
```
Comment (outcatcher) [inline src/oms/core/signer.ts:95]: You can skip the return type definition for non-exported fields

File: src/oms/core/signer.ts [hunk 6]
```diff
@@ -1,46 +1,102 @@
...
+        const stringToSign = `${algorithm}\n${amzDate}\n${credentialScope}\n${sha256(canonicalRequest).toString()}`
+        return { stringToSign, algorithm, credentialScope }
+    }
+
+    private static signString(input: SignatureInputData, dateStamp: string, stringToSign: string) {
+        const signingKey = Signature.getSignatureKey(input.secretKey, dateStamp, input.region, input.service)
+        return sha256.hmac(stringToSign, signingKey).toString()
+    }
+
+    private static generateAuthorizationHeader(algorithm: string, input: SignatureInputData, credentialScope: string, signedHeaders: string, signature: any) {
```
Comment (outcatcher) [inline src/oms/core/signer.ts:87]: Can `signature` really be `any`?

File: src/oms/core/signer.ts [hunk 7]
```diff
@@ -1,46 +1,102 @@
...
-//  HexEncode(Hash(RequestPayload))
-function canonicalRequest(config: RequestOpts): string {
-    const url = new URL(config.url, config.baseURL)
-    const queryString = url.search.substring(1)
-    const canonHeaders = Object.entries(config.headers).map(e => `${e[0].toLowerCase()}:${e[1].trim()}\n`).sort().join('')
-    const signedHeaders = Object.keys(config.headers).sort((a: string, b: string): number => (a[0].toLowerCase() < b[0].toLowerCase() ? -1 : 1)).join(';')
-    const hash = hexEncode(sha256(config.body!))
-    return `${config.method}\n${url.pathname}\n${queryString}\n${canonHeaders}\n${signedHeaders}\n${hash}`
+
+export interface SignatureInputData{
```
Comment (outcatcher) [inline src/oms/core/signer.ts:8]: Interface fields can be readonly

File: src/oms/index.ts [hunk 8]
```diff
@@ -99,13 +99,38 @@ export class Client {
...
+        let region = this.cloud.region
+        if (!region){
+            region = defaultRegion
+        }
         if (!this.cloud.auth.ak || !this.cloud.auth.sk) {
             throw Error(`Missing AK/SK: ${JSON.stringify(this.cloud.auth)}`)
         }
+        headers.append( 'content-type', 'application/json')
+
+        const authData: SignatureInputData = {
```
Comment (outcatcher) [inline src/oms/index.ts:114]: You can skip type definition to make one less import

File: src/oms/core/signer.ts [hunk 9]
```diff
@@ -1,46 +1,107 @@
...
-        config.headers.set('Authorization', canonicalRequest(config))
-        return config
+export class Signature {
+    /**
+     * Generates the signature
+     *
+     * @param {SignatureInputData} input - structure with data to be signed and keys
+     * @param {Date} currentDate - optional parameter to pass custom date
+     */
+    generateSignature(input: SignatureInputData, currentDate: Date = new Date()): SignatureOutputData {
```
Comment (lego963) [inline src/oms/core/signer.ts:35]: `currentDate: Date = new Date()` has it be here?

**Pull request closed without merging**